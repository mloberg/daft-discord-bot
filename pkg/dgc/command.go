package dgc

import "github.com/bwmarrin/discordgo"

// Command represents a Discord slash command
type Command struct {
	Name        string
	Description string
	Options     []*discordgo.ApplicationCommandOption

	Run func(s *discordgo.Session, i *discordgo.InteractionCreate) error
}
