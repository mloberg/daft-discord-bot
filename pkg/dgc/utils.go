package dgc

import "github.com/bwmarrin/discordgo"

var InteractionResponseEphemeral uint64 = 1 << 6

// GetUser returns the user that created the interaction either in a DM or guild
func GetUser(i *discordgo.InteractionCreate) *discordgo.User {
	if i.Member != nil {
		return i.Member.User
	}

	return i.User
}
