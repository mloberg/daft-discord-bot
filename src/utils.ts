export function env(key: string, _default: string): string {
    return process.env[key] ?? _default;
}
