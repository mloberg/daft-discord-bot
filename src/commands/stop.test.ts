import { Client, Guild, Message, TextChannel } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import { FriendlyError } from '../error';
import command from './stop';

const mocks = {
    react: jest.fn(),
    leave: jest.fn(),
};

jest.mock('discord.js', () => {
    return {
        Client: jest.fn(),
        Guild: jest.fn(),
        TextChannel: jest.fn(),
        Message: jest.fn().mockImplementation(() => {
            return {
                member: {
                    voice: {
                        channel: {
                            leave: mocks.leave,
                        },
                    },
                },
                react: mocks.react,
            };
        }),
    };
});

describe('_stop configuration', () => {
    it('should have basic command infomation', () => {
        expect(command.name).toEqual('stop');
        expect(command.description).toEqual('Stop the playlist');
    });

    it('should have no aliases', () => {
        expect(command.alias).toBeUndefined();
    });
});

describe('_stop', () => {
    const client = new Client();
    const guild = new Guild(client, {});
    const channel = new TextChannel(guild, {});

    beforeEach(() => {
        mocks.react.mockClear();
        mocks.leave.mockClear();
    });

    it('leaves the voice channel', async () => {
        const message = new Message(client, {}, channel);

        await command.run(message, { _: [], $0: 'stop' });

        expect(mocks.leave).toHaveBeenCalledTimes(1);
        expect(mocks.react).toBeCalledWith('ðŸŽ¶');
    });

    it('will throw an error if not in a voice channel', async () => {
        (mocked(Message) as jest.Mock).mockImplementationOnce(() => {
            return { member: { voice: { channe: null } } };
        });
        const message = new Message(client, {}, channel);

        try {
            await command.run(message, { _: [], $0: 'stop' });
            fail('expected error to be thrown');
        } catch (err) {
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('You are not in a voice channel');
        }
    });
});
