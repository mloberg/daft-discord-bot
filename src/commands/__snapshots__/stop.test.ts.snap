// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_stop is a command 1`] = `
Object {
  "description": "Stop the playlist",
  "name": "stop",
  "run": [Function],
}
`;
